eos/utils/expression-parser_TEST.cc

```diff
@@ -0,0 +1,138 @@
+using namespace eos;
+
+
+class ExpressionEvaluator
+{
+    public:
+        double visit(BinaryExpression & e)
+        {
+            BinaryExpression::func f = BinaryExpression::Method(e._op);
+
+            return f(e._lhs.accept_returning<double>(*this), e._rhs.accept_returning<double>(*this));
+        }
+
+        double visit(ConstantExpression & e)
+        {
+            return e._value;
+        }
+
+        double visit(ObservableExpression & e)
+        {
+            return e._observable->evaluate();
+        }
+
+        double visit(ObservableNameExpression &)
+        {
+            throw InternalError("Encountered ObserableNameExpression in ExpressionEvaluator::visit");
+
+            return 0.0;
+        }
+};
+
+
+struct ExpressionTest
+{
+    const std::string _input;
+
+    Expression _e;
+    bool _completed;
+
+
+    typedef std::string::const_iterator It;
+    ExpressionParser<It> p;
+
+    ExpressionTest(const std::string input) :
+        _input(input)
+    {
+        It f(_input.begin()), l(_input.end());
+        _completed = qi::phrase_parse(f, l, p, ascii::space, _e);
+    }
+};
+
+
+class ExpressionParserTest :
+    public TestCase
+{
+public:
+    ExpressionParserTest() :
+    TestCase("expression_parser_test")
+    {
+    }
+
+    virtual void run() const
+    {
+        {
+            ExpressionTest test("1+2*3");
+
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test._e.accept(printer);
+
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test._completed);
+            TEST_CHECK_EQUAL(test._e.accept_returning<double>(evaluator), 7);
+            TEST_CHECK_EQUAL(out.str(), "BinaryExpression(ConstantExpression(1) + BinaryExpression(ConstantExpression(2) * ConstantExpression(3)))");
+        }
+
+        {
+            ExpressionTest test2("{B->Dlnu::BR;l=tau} / {B->Dlnu::BR;l=mu}");
+
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test2._e.accept(printer);
+
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test2._completed);
+            TEST_CHECK_EQUAL(out.str(),
+                "BinaryExpression(ObservableNameExpression(B->Dlnu::BR;l=tau) / ObservableNameExpression(B->Dlnu::BR;l=mu))"
+                );
+
+            // Cannot evaluate expression with ObservableNameExpression objects
+            TEST_CHECK_THROWS(InternalError, test2._e.accept_returning<double>(evaluator));
+
+            Parameters p = Parameters::Defaults();
+            Kinematics k = Kinematics({{"q2_min", 0.011164}, {"q2_max", 11.62}});
+            Options o;
+            ExpressionMaker maker(p, k, o);
+            Expression evaluable_test2 = test2._e.accept_returning<Expression>(maker);
+
+            // This test is to be improved with correct kinematics and default parameters
+            TEST_CHECK_RELATIVE_ERROR(evaluable_test2.accept_returning<double>(evaluator), 0.295627626, 1e-5);
```

Comment:
Hm, this provide a circular dependence on the EOS libraries, but only when testing: ``libeosutils`` must be built before ``libeosbdecays.so``, but the tests of ``libeosutils`` depend on ``libeosbdecays``. Since it works, let's be happy and leave it as is.